### PR TITLE
File observer fixes

### PR DESCRIPTION
**Pull request by mfojtik**
* First commit is adding mechanism for callers to ensure the initial sync passed and the fileobserver see and observing changes on files successfully
* Second commit is critical as when the initial content of file is empty, the reactor for the file is never added
* Third commit is adding much more unit test coverage to ensure all corner cases are handled

There is one TODO in unit test when the initial content is empty and the file content is empty (but exists), the observer will trigger "Created" action, even when the file existed. This should be handled as follow up issue.

### CODE REVIEW

File: pkg/controller/fileobserver/observer_polling.go [hunk 1]
```diff
@@ -20,6 +20,12 @@ type pollingObserver struct {
 	files    map[string]string
 
 	reactorsMutex sync.RWMutex
+	hasSynced     bool
+}
+
+// HasSynced indicates that the observer synced all observed files at least once.
+func (o *pollingObserver) HasSynced() bool {
+	return o.hasSynced
```
Comment (sttts) [inline pkg/controller/fileobserver/observer_polling.go]: must this be go routine safe?

Comment (mfojtik) [reply pkg/controller/fileobserver/observer_polling.go]: i thought so as well, but aren't booleans safe in go?

Comment (mfojtik) [reply pkg/controller/fileobserver/observer_polling.go]: and I was wrong ;-) the data race in unit test :-)

Comment (sttts) [reply pkg/controller/fileobserver/observer_polling.go]: No, they are not.

Comment (mfojtik) [reply pkg/controller/fileobserver/observer_polling.go]: fixed by mutex

**Comment by mfojtik** [comment]
/hold

todo must be fixed

File: pkg/controller/fileobserver/observer_polling.go [hunk 2]
```diff
@@ -74,28 +86,42 @@ func (o *pollingObserver) processReactors(stopCh <-chan struct{}) {
...
+			klog.Infof("Observed change: file:%s (current: %q, lastKnown: %q)", filename, currentFileState.hash, lastKnownFileState.hash)
 
 			for i := range reactors {
-				action := FileModified
-				switch {
-				case len(lastKnownHash) == 0:
+				action := NoAction
+
+				// do nothing if file does not exists
+				if !lastKnownFileState.exists && !currentFileState.exists {
```
Comment (sttts) [inline pkg/controller/fileobserver/observer_polling.go]: all these cases would make a nice switch statement (again).

**Comment by sttts** [comment]
Lgtm. Just some nit.

File: pkg/controller/fileobserver/observer_polling.go (hunk 2 above)
Comment (mfojtik) [reply pkg/controller/fileobserver/observer_polling.go]: @sttts i found the switch more confusing, but I can follow up? :-) at least I put comments on what each case mean
